File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.py
from sys import stdin

n = int(stdin.readline())
numbers = list(map(int, stdin.readline().split()))
op_cnt = list(map(int, stdin.readline().split()))
op = {'+': op_cnt[0], '-': op_cnt[1], '*': op_cnt[2], '/': op_cnt[3]}
max_value = -int(1e9)
min_value = int(1e9)


def dfs(depth: int, total: int) -> None:
    global numbers, max_value, min_value, op, n
    if depth == n:
        max_value = max_value if max_value > total else total
        min_value = min_value if min_value < total else total
        return
    if op['+']:
        op['+'] -= 1
        dfs(depth + 1, total + numbers[depth])
        op['+'] += 1
    if op['-']:
        op['-'] -= 1
        dfs(depth + 1, total - numbers[depth])
        op['-'] += 1
    if op['*']:
        op['*'] -= 1
        dfs(depth + 1, total * numbers[depth])
        op['*'] += 1
    if op['/']:
        op['/'] -= 1
        dfs(depth + 1, int(total / numbers[depth]))
        op['/'] += 1


dfs(1, numbers[0])
print(max_value)
print(min_value)

File: 백준/Silver/14888. 연산자 끼워넣기/연산자 끼워넣기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

class Main{

    private static int[] count, numbers;
    private static int n, minValue, maxValue;

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());
        numbers = new int[n];
        count = new int[4];

        StringTokenizer tokenizer = new StringTokenizer(br.readLine());
        for(int i=0 ; i<n ;i++){
            numbers[i] = Integer.parseInt(tokenizer.nextToken());
        }

        tokenizer = new StringTokenizer(br.readLine());
        for(int i=0 ; i<4 ; i++){
            count[i] = Integer.parseInt(tokenizer.nextToken());
        }

        minValue = Integer.MAX_VALUE;
        maxValue = Integer.MIN_VALUE;
        backtracking(1, numbers[0]);

        System.out.println(maxValue);
        System.out.println(minValue);
    }

    private static void backtracking(int depth, int total){
        if(depth == n){
            minValue = Math.min(minValue, total);
            maxValue = Math.max(maxValue, total);
            return;
        }

        if(count[0] > 0){
            count[0] -= 1;
            backtracking(depth + 1, total + numbers[depth]);
            count[0] += 1;
        }
        if(count[1] > 0){
            count[1] -= 1;
            backtracking(depth + 1, total - numbers[depth]);
            count[1] += 1;
        }
        if(count[2] > 0){
            count[2] -= 1;
            backtracking(depth + 1, total * numbers[depth]);
            count[2] += 1;
        }
        if(count[3] > 0){
            count[3] -= 1;
            backtracking(depth + 1, total / numbers[depth]);
            count[3] += 1;
        }
    }
}